### PR TITLE
repl bug fixes

### LINKED ISSUE #14438: The REPL crashes with UTF-8 entries and rsearch
At REPL do: 
```
ù
ùù
Ctrl-r
ù
Ctrl-r
Ctrl-s
Ctrl-s
```
Then Julia crashes with an error including in particular: "type SearchState has no field p".
This problem disappears with #9352, but I have no idea if that PR only buries the bug without actually solving it.
EDIT: now that the error doesn't show up anymore in master (not necessarily meaning that the bug is solved), here is the full error text:
```
WARNING: Caught an exception in the keymap:
ERROR: UnicodeError: invalid character index
ERROR: type SearchState has no field p
 in run_interface at ./LineEdit.jl:1610
 in run_frontend at ./REPL.jl:864
 in run_repl at ./REPL.jl:167
 in _start at ./client.jl:419
```

### PR DESCRIPTION
**Pull request by vtjnash**
I've gone through many of the issues tagged "REPL" and tried to knock out a lot of these known lingering bugs. 

(I also attempted to fix #14438, but since it seems that can only be triggered by some error state, I had no way to test it)

### CODE REVIEW

File: test/repl.jl [hunk 1]
```diff
@@ -159,8 +159,8 @@ end
 function AddCustomMode(repl)
     # Custom REPL mode tests
     foobar_mode = LineEdit.Prompt("Test";
-        prompt_prefix="\e[38;5;166m",
-        prompt_suffix=Base.text_colors[:white],
+        prompt_prefix="\e[38;5;166mΠ",
```
Comment (Keno) [inline test/repl.jl]: What's `Π` for? The intention was to not have any printable characters in here.

Comment (vtjnash) [inline test/repl.jl]: for seeing if anything broke if random unicode characters were inserted (although it looks like I forgot to insert them into the prompt itself as I had also intended :/ )

**Comment by vtjnash** [comment]
can i merge this? i'm really missing the paste off-by-one bug fix everytime i work on something else :stuck_out_tongue:

**Comment by Keno** [comment]
Did you want to change the test back to remove the random Unicode character?

**Comment by vtjnash** [comment]
i wasn't intending to. even though the usage is undocumented as being incorrect, the code should be able to handle it, so it seems like a valid thing to have in the tests.

**Comment by Keno** [comment]
It doesn't make any sense. It would be fine to break it, so it shouldn't be in the test.

**Pull request merged**